command/agent/auth/auth.go

```diff
@@ -254,9 +255,22 @@ func (ah *AuthHandler) Run(ctx context.Context, am AuthMethod) error {
 		}
 
 		// This should only happen if there's no preloaded token (regular auto-auth login)
-		//  or if a preloaded token has expired and is now switching to auto-auth.
+		// or if a preloaded token has expired and is now switching to auto-auth.
 		if secret.Auth == nil {
-			secret, err = clientToUse.Logical().WriteWithContext(ctx, path, data)
+			isTokenFileMethod = path == "auth/token/lookup-self"
```

Comment:
I would have more preferably checked if the auth method was in fact `TokenFileMethod`, but doing so would have caused a circular dependency.

Reply:
You could add a new method to the AuthMethod interface, e.g. IsStaticCredentials.  Or equivalently, you could introduce a new sub-interface, like

```
type TokenSource interface {
	Authenticate(context.Context, *api.Client) (string, http.Header, map[string]interface{}, error)
}

type AuthMethod interface {
        TokenSource
	NewCreds() chan struct{}
	CredSuccess()
	Shutdown()
}
```

Going further, we could have Authenticate do the request itself, instead of returning a path and header and data body, meaning

```
type TokenSource interface {
	Authenticate(context.Context, *api.Client) (*api.Secret, error)
}
```

This seems like a more natural and less cumbersome approach than the current Authenticate method signature.

Reply:
I did think about adding to the `AuthMethod` interface (a few times, actually, I had also toyed with `AuthMethod` having a new method for getting the important bits out of the returned `*api.Secret`, as well as this use-case), but it seemed strange, in that I'd expect that only `tokenFileMethod` would ever be different. It's unclear which is cleaner, to me. If you prefer extending `AuthMethod` in this case, I'm happy to change approach!

Reply:
I don't have a problem with your current approach.  My proposal is more disruptive and not that much better, so if it doesn't speak to you feel free to ignore it.

Reply:
Just to add my thoughts to this quickly: 

I don't have a concrete suggestion as such to offer right now (sorry), but it feels like we're maybe doing something that's certainly workable, but perhaps doesn't follow existing ways we might be encapsulating (whatever the Go term is, modularisation?) the behaviour etc. and it bleeds into a place it might not need to be. 

In the absence of an example to suggest an immediate alternative, I fall back to thinking, are we showing developer empathy, would this code be easy to maintain for the team, should we be trying to refactor it while we're in the code for this work item?

I ran the `gocyclo` and `gocognit` tools for the `auth.go` file and I think we're in danger of making code even harder to understand :( 

`auth.go/Run`:
* `cyclo`: 54 => 67
* `cognit`: 136 => 184

As long as we're making a balanced judgement on this overall, I'm happy but wanted to share in case it helped.  :)

Reply:
Yeah, an unfortunate implication is this feature is that it adds unavoidable complexity to Authenticate, as it needs to handle the case where:
- It authenticates
- (new!) It validates a previous authentication

There's downsides to either approach, but the reason I took the one I did was that I believe this approach minimizes complexity when adding new (or maintaining old) auto-auth methods to agent, which I believe is something we'll need to do/maintain more than this method here. I could be wrong, to be clear, but I wanted to be a bit clearer as to why I thought this was the best place to shoulder the complexity burden, and I'm still happy to change if folks feel like this isn't the best place to bear the complexity.

I appreciate you bringing it up, though, it's good to bring to the forefront of all of our minds, and think about.